client/src/main.rs

```diff
@@ -1,12 +1,218 @@
+    info!("Sent command: {:?}", command);
+}
+
+// mw for mouse wheel, mm for mouse motion
+fn handle_mouse_input(
+    input: DeviceEvent,
+    protocol: &mut Protocol,
+    mm_buffer: &mut Queue<DeviceEvent>,
+    mw_buffer: &mut Queue<DeviceEvent>,
+) {
+    match input {
+        DeviceEvent::MouseMotion { .. } => {
+            mm_buffer.add(input).expect("adding to mm_buffer failed \n");
+        }
+        DeviceEvent::MouseWheel { .. } => {
+            mw_buffer.add(input).expect("adding to mw_buffer failed \n");
+        }
+        // what's that possibly for?
+        DeviceEvent::Button { .. } => {
+            // if we receive those button events, then should send right away with protocol
+            let mut command: Option<Command> = None;
+        }
+        _ => {}
+    }
+}
 
-use common::add;
-use client::run;
 fn main() {
-    println!("Client: Hello, world!");
+    env_logger::init();
+    let (tx, rx) = mpsc::channel::<UserInput>();
+    let game_state = Arc::new(Mutex::new(GameState::default()));
+
+    let server_details = DEFAULT_SERVER_ADDR;
+    let dest: SocketAddr = server_details.parse().expect("server details parse fails");
+
+    let mut protocol = Protocol::connect(dest).unwrap();
+
+    let mut client_id: u32 = 0;
+
+    let mut event_loop = PlayerLoop::new(tx, client_id);
+
+    thread::spawn(move || {
+        // TODO: Initial Connection to get current client id
+        while let Ok(msg) = protocol.read_message::<Message>() {
+            match msg {
+                Message {
+                    host_role: HostRole::Server,
+                    payload,
+                    ..
+                } => match payload {
+                    Payload::Init(incoming_id) => {
+                        client_id = incoming_id;
+                        info!("Connection established ");
+                        break;
+                    }
+                    _ => {
+                        error!("first contact with server is not init \n");
+                    }
+                },
+                _ => {
+                    error!("first contact with server is not init \n");
+                }
+            }
+        }
+
+        let mut mm_tot_dx;
+        let mut mm_tot_dy;
+        let mut mw_tot_line_dx;
+        let mut mw_tot_line_dy;
+        let mut mw_tot_pixel_dx;
+        let mut mw_tot_pixel_dy;
+
+        let mut mouse_motion_buf = Queue::new();
+        let mut mouse_wheel_buf = Queue::new();
+        let mut sample_start_time = Instant::now();
+
+        loop {
+            while let Ok(user_input) = rx.recv() {
+                match user_input.input {
+                    Inputs::Keyboard(input) => {
+                        handle_keyboard_input(input, &mut protocol);
+                        // handle keyboard event
+                        // check for new state & update local game state
+                        while let Ok(msg) = protocol.read_message::<Message>() {
```

Comment:
why does the input processing loop block on receiving state syncs? I think this will make the client tick rate >= server tick rate.